In [1]:
import pandas as pd
import numpy as np

In [2]:
diary_data = pd.read_stata('dati/td_ita.dta')

## MISSING VALUES IN DIARY DATA

**Goal**: remove respondents that have too many missing values. What does it mean 'too many'? Still to be decided.

### Question: how many participants are there?

In [3]:
len(list(diary_data['id'].unique()))

241

### Question: how many lines for a single person?

In [22]:
group_ids = diary_data.groupby('id')
list(group_ids.apply(len)) # len of the list is 241

[1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,
 1114,

In [5]:
241*1114

268474

We get the exact number of lines of the original df.

Every line is a record that is done every half-an-hour during each day. How many days are there for each pp?

In [6]:
diary_data['day_not_str'] = diary_data['date_not'].dt.strftime('%Y-%m-%d')

In [7]:
diary_data.head() #ok

pilot   id            date_not         datein_ques  \
0  wenetItaly  0.0 2020-11-13 00:00:00 2020-11-13 00:00:00   
1  wenetItaly  0.0 2020-11-13 00:30:00 2020-11-13 00:30:00   
2  wenetItaly  0.0 2020-11-13 01:00:00 2020-11-13 01:00:00   
3  wenetItaly  0.0 2020-11-13 01:30:00 2020-11-13 01:30:00   
4  wenetItaly  0.0 2020-11-13 02:00:00 2020-11-13 02:00:00   

          datein_answ          first2w    week  YY_not  MM_not  DD_not  ...  \
0 2020-11-13 00:00:00  First two weeks  Friday  2020.0    11.0    13.0  ...   
1 2020-11-13 00:30:00  First two weeks  Friday  2020.0    11.0    13.0  ...   
2 2020-11-13 01:00:00  First two weeks  Friday  2020.0    11.0    13.0  ...   
3 2020-11-13 01:30:00  First two weeks  Friday  2020.0    11.0    13.0  ...   
4 2020-11-13 02:00:00  First two weeks  Friday  2020.0    11.0    13.0  ...   

   c6_18  c6_19 c6_20 c6_21 c6_22 c6_23 c6_24 c6_25 c6_26 day_not_str  
0    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-13  
1    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-13  
2    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-13  
3    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-13  
4    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-13  

[5 rows x 102 columns]

In [8]:
print(list(diary_data['day_not_str'].unique()))
len(list(diary_data['day_not_str'].unique()))

['2020-11-13', '2020-11-14', '2020-11-15', '2020-11-16', '2020-11-17', '2020-11-18', '2020-11-19', '2020-11-20', '2020-11-21', '2020-11-22', '2020-11-23', '2020-11-24', '2020-11-25', '2020-11-26', '2020-11-27', '2020-11-28', '2020-11-29', '2020-11-30', '2020-12-01', '2020-12-02', '2020-12-03', '2020-12-04', '2020-12-05', '2020-12-06', '2020-12-07', '2020-12-08', '2020-12-09', '2020-12-10', '2020-12-11']


29

In [16]:
29*48 

1392

It is not 1114... so??? Let's try to group by days and check the length of each group.

In [9]:
group_days = diary_data.groupby('day_not_str')
print(group_days.apply(len))

day_not_str
2020-11-13    11568
2020-11-14    11568
2020-11-15    11568
2020-11-16    11568
2020-11-17    11568
2020-11-18    11568
2020-11-19    11568
2020-11-20    11568
2020-11-21    11568
2020-11-22    11568
2020-11-23    11568
2020-11-24    11568
2020-11-25    11568
2020-11-26    11568
2020-11-27    11568
2020-11-28    11568
2020-11-29    11568
2020-11-30     8194
2020-12-01     5784
2020-12-02     5784
2020-12-03     5784
2020-12-04     5784
2020-12-05     5784
2020-12-06     5784
2020-12-07     5784
2020-12-08     5784
2020-12-09     5784
2020-12-10     5784
2020-12-11     5784
dtype: int64


What does this mean??? We will try to find out I guess... Let's explore the day with 8194 records.

In [16]:
 group_days.get_group('2020-11-30').head()

pilot   id            date_not         datein_ques  \
816  wenetItaly  0.0 2020-11-30 00:00:00 2020-11-30 00:12:36   
817  wenetItaly  0.0 2020-11-30 00:30:00 2020-11-30 00:35:06   
818  wenetItaly  0.0 2020-11-30 01:00:00 2020-11-30 08:53:31   
819  wenetItaly  0.0 2020-11-30 01:30:00 2020-11-30 08:53:31   
820  wenetItaly  0.0 2020-11-30 02:00:00 2020-11-30 08:53:32   

            datein_answ          first2w    week  YY_not  MM_not  DD_not  ...  \
816 2020-11-30 00:14:13  First two weeks  Monday  2020.0    11.0    30.0  ...   
817 2020-11-30 09:52:42  First two weeks  Monday  2020.0    11.0    30.0  ...   
818 2020-11-30 09:52:48  First two weeks  Monday  2020.0    11.0    30.0  ...   
819 2020-11-30 09:52:53  First two weeks  Monday  2020.0    11.0    30.0  ...   
820 2020-11-30 09:52:58  First two weeks  Monday  2020.0    11.0    30.0  ...   

     c6_18  c6_19 c6_20 c6_21 c6_22 c6_23 c6_24 c6_25 c6_26 day_not_str  
816    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-30  
817    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-30  
818    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-30  
819    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-30  
820    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  2020-11-30  

[5 rows x 102 columns]

notififcations every half an hour.

In [17]:
 group_days.get_group('2020-11-30').tail()

pilot     id            date_not         datein_ques  \
268205  wenetItaly  266.0 2020-11-30 19:00:00 2020-11-30 19:00:00   
268206  wenetItaly  266.0 2020-11-30 20:00:00 2020-11-30 20:00:00   
268207  wenetItaly  266.0 2020-11-30 21:00:00 2020-11-30 21:00:00   
268208  wenetItaly  266.0 2020-11-30 22:00:00 2020-11-30 22:00:00   
268209  wenetItaly  266.0 2020-11-30 23:00:00 2020-11-30 23:00:00   

               datein_answ          first2w    week  YY_not  MM_not  DD_not  \
268205 2020-11-30 22:22:42  First two weeks  Monday  2020.0    11.0    30.0   
268206 2020-11-30 22:22:52  First two weeks  Monday  2020.0    11.0    30.0   
268207 2020-11-30 22:23:03  First two weeks  Monday  2020.0    11.0    30.0   
268208 2020-11-30 22:23:26  First two weeks  Monday  2020.0    11.0    30.0   
268209 2020-12-01 11:00:00  First two weeks  Monday  2020.0    11.0    30.0   

        ...  c6_18  c6_19 c6_20 c6_21 c6_22 c6_23 c6_24 c6_25 c6_26  \
268205  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268206  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268207  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268208  ...     No     No    No    No    No    No    No    No   Yes   
268209  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

       day_not_str  
268205  2020-11-30  
268206  2020-11-30  
268207  2020-11-30  
268208  2020-11-30  
268209  2020-11-30  

[5 rows x 102 columns]

Notifications **every hour!**

Maybe the fact that some notifications are every hour happens in general in some days?

In [19]:
diary_data.tail()

pilot     id            date_not         datein_ques  \
268469  wenetItaly  266.0 2020-12-11 19:00:00 2020-12-11 19:00:00   
268470  wenetItaly  266.0 2020-12-11 20:00:00 2020-12-11 20:06:58   
268471  wenetItaly  266.0 2020-12-11 21:00:00 2020-12-11 21:00:00   
268472  wenetItaly  266.0 2020-12-11 22:00:00 2020-12-11 22:00:00   
268473  wenetItaly  266.0 2020-12-11 23:00:00 2020-12-11 23:00:00   

               datein_answ           first2w    week  YY_not  MM_not  DD_not  \
268469 2020-12-11 21:38:15  Second two weeks  Friday  2020.0    12.0    11.0   
268470 2020-12-11 21:38:24  Second two weeks  Friday  2020.0    12.0    11.0   
268471 2020-12-11 21:38:37  Second two weeks  Friday  2020.0    12.0    11.0   
268472 2020-12-12 00:14:37  Second two weeks  Friday  2020.0    12.0    11.0   
268473 2020-12-12 00:14:54  Second two weeks  Friday  2020.0    12.0    11.0   

        ...  c6_18  c6_19 c6_20 c6_21 c6_22 c6_23 c6_24 c6_25 c6_26  \
268469  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268470  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268471  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268472  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268473  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

       day_not_str  
268469  2020-12-11  
268470  2020-12-11  
268471  2020-12-11  
268472  2020-12-11  
268473  2020-12-11  

[5 rows x 102 columns]

It seems to be happening also in the last day of data collection. Let's dig deeper into this and explore the last pp during the last day.How many rows will it have?

In [31]:
filt = (diary_data['id'] == 266.0) & (diary_data['day_not_str'] == '2020-12-11')
diary_data.loc[filt]

pilot     id            date_not         datein_ques  \
268450  wenetItaly  266.0 2020-12-11 00:00:00 2020-12-11 00:00:00   
268451  wenetItaly  266.0 2020-12-11 01:00:00 2020-12-11 01:00:00   
268452  wenetItaly  266.0 2020-12-11 02:00:00 2020-12-11 02:00:00   
268453  wenetItaly  266.0 2020-12-11 03:00:00 2020-12-11 03:00:00   
268454  wenetItaly  266.0 2020-12-11 04:00:00 2020-12-11 04:00:00   
268455  wenetItaly  266.0 2020-12-11 05:00:00 2020-12-11 05:00:00   
268456  wenetItaly  266.0 2020-12-11 06:00:00 2020-12-11 06:00:00   
268457  wenetItaly  266.0 2020-12-11 07:00:00 2020-12-11 07:00:00   
268458  wenetItaly  266.0 2020-12-11 08:00:00 2020-12-11 08:00:01   
268459  wenetItaly  266.0 2020-12-11 09:00:00 2020-12-11 09:00:01   
268460  wenetItaly  266.0 2020-12-11 10:00:00 2020-12-11 10:00:01   
268461  wenetItaly  266.0 2020-12-11 11:00:00 2020-12-11 11:00:01   
268462  wenetItaly  266.0 2020-12-11 12:00:00 2020-12-11 12:00:00   
268463  wenetItaly  266.0 2020-12-11 13:00:00 2020-12-11 13:00:00   
268464  wenetItaly  266.0 2020-12-11 14:00:00 2020-12-11 14:00:00   
268465  wenetItaly  266.0 2020-12-11 15:00:00 2020-12-11 15:00:00   
268466  wenetItaly  266.0 2020-12-11 16:00:00 2020-12-11 16:00:00   
268467  wenetItaly  266.0 2020-12-11 17:00:00 2020-12-11 17:00:00   
268468  wenetItaly  266.0 2020-12-11 18:00:00 2020-12-11 18:00:00   
268469  wenetItaly  266.0 2020-12-11 19:00:00 2020-12-11 19:00:00   
268470  wenetItaly  266.0 2020-12-11 20:00:00 2020-12-11 20:06:58   
268471  wenetItaly  266.0 2020-12-11 21:00:00 2020-12-11 21:00:00   
268472  wenetItaly  266.0 2020-12-11 22:00:00 2020-12-11 22:00:00   
268473  wenetItaly  266.0 2020-12-11 23:00:00 2020-12-11 23:00:00   

               datein_answ           first2w    week  YY_not  MM_not  DD_not  \
268450 2020-12-11 01:39:39  Second two weeks  Friday  2020.0    12.0    11.0   
268451 2020-12-11 01:39:47  Second two weeks  Friday  2020.0    12.0    11.0   
268452 2020-12-11 02:39:26  Second two weeks  Friday  2020.0    12.0    11.0   
268453 2020-12-11 03:04:27  Second two weeks  Friday  2020.0    12.0    11.0   
268454 2020-12-11 13:15:46  Second two weeks  Friday  2020.0    12.0    11.0   
268455 2020-12-11 13:16:08  Second two weeks  Friday  2020.0    12.0    11.0   
268456 2020-12-11 13:16:14  Second two weeks  Friday  2020.0    12.0    11.0   
268457 2020-12-11 13:16:20  Second two weeks  Friday  2020.0    12.0    11.0   
268458 2020-12-11 13:16:40  Second two weeks  Friday  2020.0    12.0    11.0   
268459 2020-12-11 13:16:46  Second two weeks  Friday  2020.0    12.0    11.0   
268460 2020-12-11 13:16:52  Second two weeks  Friday  2020.0    12.0    11.0   
268461 2020-12-11 13:16:57  Second two weeks  Friday  2020.0    12.0    11.0   
268462 2020-12-11 21:36:57  Second two weeks  Friday  2020.0    12.0    11.0   
268463 2020-12-11 21:37:09  Second two weeks  Friday  2020.0    12.0    11.0   
268464 2020-12-11 21:37:23  Second two weeks  Friday  2020.0    12.0    11.0   
268465 2020-12-11 21:37:35  Second two weeks  Friday  2020.0    12.0    11.0   
268466 2020-12-11 21:37:44  Second two weeks  Friday  2020.0    12.0    11.0   
268467 2020-12-11 21:37:59  Second two weeks  Friday  2020.0    12.0    11.0   
268468 2020-12-11 21:38:07  Second two weeks  Friday  2020.0    12.0    11.0   
268469 2020-12-11 21:38:15  Second two weeks  Friday  2020.0    12.0    11.0   
268470 2020-12-11 21:38:24  Second two weeks  Friday  2020.0    12.0    11.0   
268471 2020-12-11 21:38:37  Second two weeks  Friday  2020.0    12.0    11.0   
268472 2020-12-12 00:14:37  Second two weeks  Friday  2020.0    12.0    11.0   
268473 2020-12-12 00:14:54  Second two weeks  Friday  2020.0    12.0    11.0   

        ...  c6_18  c6_19 c6_20 c6_21 c6_22 c6_23 c6_24 c6_25 c6_26  \
268450  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268451  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
268452  ...    NaN    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
26

Just 24! So on this day this partecipant had one notification every hour. But still, for this partecipant we have 1114 rows in total. Can 1114 be divided by 24?

In [32]:
1114/24

46.416666666666664

Nope... so what? Still needed to be figured out.

In [41]:
group_ids['day_not_str'].value_counts().head(29) # is this the pattern??

id   day_not_str
0.0  2020-11-13     48
     2020-11-14     48
     2020-11-22     48
     2020-11-28     48
     2020-11-27     48
     2020-11-26     48
     2020-11-25     48
     2020-11-24     48
     2020-11-23     48
     2020-11-21     48
     2020-11-20     48
     2020-11-19     48
     2020-11-18     48
     2020-11-17     48
     2020-11-16     48
     2020-11-15     48
     2020-11-29     48
     2020-11-30     34
     2020-12-01     24
     2020-12-09     24
     2020-12-10     24
     2020-12-11     24
     2020-12-08     24
     2020-12-02     24
     2020-12-03     24
     2020-12-04     24
     2020-12-05     24
     2020-12-06     24
     2020-12-07     24
Name: count, dtype: int64

In [40]:
(24*11)+34+(48*17)

1114

Pattern finally found!